### PR TITLE
Add error message if the return type of a step function doesn't match the expected type

### PR DESCRIPTION
**Pull request by schustmi**
## Pre-requisites
Please ensure you have done the following:
- [x] I have read the **CONTRIBUTING.md** document.
- [x] If my change requires a change to the documentation, I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Describe changes
Add error message if the return type of a step function doesn't match the expected type

### CODE REVIEW

File: src/zenml/steps/utils.py [hunk 1]
```diff
@@ -410,19 +410,26 @@ def Do(
         return_type: Type[Any] = spec.annotations.get("return", None)
         if return_type is not None:
             if isinstance(return_type, Output):
-                # Resolve named (and multi-) outputs.
```
Comment (htahir1) [inline src/zenml/steps/utils.py:413]: Are you sure we can jut remove this block of code? What is the reasoning for removing this entire logic? Seems unrelated to the error message

File: src/zenml/steps/utils.py [hunk 2]
```diff
@@ -410,19 +410,26 @@ def Do(
...
+            # if there is only one output annotation (either directly specified
+            # or contained in an `Output` tuple) we treat the step function
+            # return value as the return for that output
+            if len(output_annotations) == 1:
+                return_values = [return_values]
+
+            for return_value, (output_name, output_type) in zip(
+                return_values, output_annotations
+            ):
+                if not isinstance(return_value, output_type):
```
Comment (htahir1) [inline src/zenml/steps/utils.py:426]: I was wondering whether we should do this at the materializer base class? The reason is that we might want to extend this in the future to also allow sub-classes etc

**Review by htahir1 — COMMENTED** [review]
Left some thoughts on the matter

File: src/zenml/steps/utils.py (hunk 1 above)
Comment (schustmi) [reply src/zenml/steps/utils.py:413]: I think all the logic should still be the same, I just unified it into one for-loop instead of two so we don't have to duplicate everything

File: src/zenml/steps/utils.py (hunk 2 above)
Comment (schustmi) [reply src/zenml/steps/utils.py:426]: I thought about that too, but the materializer doesn't actually have access to the type annotation and can in some cases handle multiple unrelated objects (a `List` and an `int` for example). I also think `isinstance` already handles subclasses sufficiently: As long as the returned object is of the specified class or any subclass, it will pass.

Comment (htahir1) [reply src/zenml/steps/utils.py:426]: Ok!

File: src/zenml/steps/utils.py (hunk 1 above)
Comment (htahir1) [reply src/zenml/steps/utils.py:413]: OK had to zoom out into the file to see whats up. Much cleaner now, ill change my review!

**Review by htahir1 — APPROVED** [review]
LGTM!

File: tests/conftest.py [hunk 3]
```diff
@@ -130,7 +130,7 @@ def one_step_pipeline():
 
     @pipeline
     def _pipeline(step_):
-        pass
+        step_()
```
Comment (jwwwb) [inline tests/conftest.py:133]: For my own understanding: why are these changes needed here?

**Review by jwwwb — APPROVED** [review]

File: tests/conftest.py (hunk 3 above)
Comment (schustmi) [reply tests/conftest.py:133]: 'Calling' the step inside the pipeline usually is necessary to connect its inputs/outputs to other steps, and at the same time also generates the tfx component that we need to run it during a pipeline.

Comment (schustmi) [reply tests/conftest.py:133]: See `BaseStep.__call__`. If you forget to do that right now you'd get an error message saying that a step hasn't been called yet, basically reminding you to call that inside the pipeline. Not sure if we'd want to lift that requirement and run steps without any input artifacts automatically if they're passed to a pipeline, @htahir1 what do you think?

Comment (htahir1) [reply tests/conftest.py:133]: You mean steps that are disconnected from everything else? I'm not sure there is a requirement for that tbh. For now I'd stray away from adding functionality not asked for
